exercises/01.compartmentalisation/js.cc

```diff
@@ -0,0 +1,568 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+#include "cheri.hh"
+#include "compartment-macros.h"
+#include "microvium/microvium.h"
+#include <compartment.h>
+#include <cstdint>
+#include <cstdlib>
+#include <debug.hh>
+#include <functional>
+#include <magic_enum/magic_enum.hpp>
+#include <microvium/microvium.h>
+#include <riscvreg.h>
+#include <tuple>
+#include <type_traits>
+#include <vector>
+
+/// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<true, "JavaScript compartment">;
+using CHERI::Capability;
+
+/**
+ * Code related to storing a secret value.  The attacker should try to leak
+ * this.
+ */
+namespace
+{
+	/// The secret value.
+	int32_t secret;
+
+	/**
+	 * Set the secret the first time.
+	 *
+	 * This blocks until there is a character available on the UART.  This is
+	 * *not* a good way of getting entropy, but it's the only one that we have
+	 * in the simulator.
+	 */
+	void set_secret()
+	{
+		auto uart = MMIO_CAPABILITY(Uart, uart);
+		while (!uart->can_read()) {}
+		secret = static_cast<int32_t>(rdcycle64());
+		Debug::log("Secret stored at {}", &secret);
+	}
+
+	/**
+	 * Report the current value of the secret and pick a new one.
+	 */
+	void check_secret(int32_t guess)
+	{
+		Debug::log("Secret was {}, you guessed {}.", secret, guess);
+		if (guess == secret)
+		{
+			Debug::log("CONGRATULATIONS! You correctly leaked the secret!");
+		}
+		secret = static_cast<int32_t>(rdcycle64());
+	}
+} // namespace
+
+/**
+ * Code related to the JavaScript interpreter.
+ */
+namespace
+{
+	/**
+	 * Constants for functions exposed to JavaScript->C++ FFI
+	 *
+	 * The values here must match the ones used in cheri.js.
+	 */
+	enum Exports : mvm_HostFunctionID
+	{
+		Print = 1,
+		Move,
+		LoadCapability,
+		LoadInt,
+		Store,
+		GetAddress,
+		SetAddress,
+		GetBase,
+		GetLength,
+		GetPermissions,
+		CheckSecret,
+	};
+
+	/// Constant for the say_hello function exposed to C++->JavaScript FFI
+	static constexpr mvm_VMExportID ExportRun = 1234;
+
+	/**
+	 * Type used for the set of capabilities that JavaScript has complete
+	 * control over.
+	 */
+	using VMRegisterState = std::array<void *, 8>;
```

Comment:
This name is a little confusing since there are two layers of virtual machines here: JS and the in-JS capability machine.